extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java

```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
   }
 
   @Override
-  public CloseableIterator<InputRow> read()
+  public CloseableIterator<InputRow> read() throws IOException
   {
-    Supplier<Iterator<InputRow>> supplier = Suppliers.memoize(() -> readAsList().iterator());
-    return CloseableIterators.withEmptyBaggage(new Iterator<InputRow>() {
-      @Override
-      public boolean hasNext()
-      {
-        return supplier.get().hasNext();
-      }
-      @Override
-      public InputRow next()
-      {
-        return supplier.get().next();
+    return newReader(whichReader()).read();
+  }
+
+  class OpenCensusProtobufReaderInternal implements InputEntityReader
+  {
+    @Override
+    public CloseableIterator<InputRow> read()
+    {
+      return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+    }
+
+    @Override
+    public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
+    {
+      return OpenCensusProtobufReader.this.sample();
+    }
+  }
+
+  public InputEntityReader newReader(ProtobufReader which)
+  {
+    switch(which) {
+      case OPEN_TELEMETRY:
+        return new OpenTelemetryMetricsProtobufReader(
+            dimensionsSpec,
+            source,
+            metricDimension,
+            valueDimension,
+            metricLabelPrefix,
+            resourceLabelPrefix
+        );
+      case OPEN_CENSUS:
+      default:
+        return new OpenCensusProtobufReaderInternal();
```

Comment:
rather than creating an internal reader for this particular case, what would you think of keeping this code path more symmetrical by moving the code that handles the format switching to a new class (e.g. `HybridProtobufReader`), and then have that call either `new OpenTelemetryMetricsProtobufReader(...)` or `new OpenCensusProtobufReader(...)` so we don't pollute the OpenCensus reader code with otel-specific handling, and keep the Otel and OC readers clean counterparts of each-other.